pkg/hs/urbit-king/lib/Urbit/Vere/Http/Server.hs

```diff
@@ -396,41 +416,90 @@ app env sId liv plan which req respond =
     an exception.
 -}
-openPort :: HasLogFunc e => Bool -> [W.Port] -> RIO e (W.Port, Net.Socket)
-openPort isFake = go
-  where
-    go = \case
-        []   -> io W.openFreePort
-        x:xs -> io (tryOpen x) >>= \case
-                    Left (err∷IOError) -> do
-                        logWarn (display ("Failed to open port " <> tshow x))
-                        logWarn (display (tshow err))
-                        go xs
-                    Right ps -> do
-                        logTrace (display ("Opening port " <> tshow (fst ps)))
-                        pure ps
-
-    bindTo = if isFake then "127.0.0.1" else "0.0.0.0"
-
-    bindListenPort ∷ W.Port → Net.Socket → IO Net.PortNumber
-    bindListenPort por sok = do
-        bindAddr <- Net.getAddrInfo Nothing (Just bindTo) Nothing >>= \case
-                        []  -> error "this should never happen."
-                        x:_ -> pure (Net.addrAddress x)
-
-        Net.bind sok bindAddr
-        Net.listen sok 1
-        Net.socketPort sok
-
-    --  `inet_addr`, `bind`, and `listen` all throw `IOError` if they fail.
-    tryOpen ∷ W.Port → IO (Either IOError (W.Port, Net.Socket))
-    tryOpen por = do
-        sok <- Net.socket Net.AF_INET Net.Stream Net.defaultProtocol
-        try (bindListenPort por sok) >>= \case
-            Left exn  -> Net.close sok $> Left exn
-            Right por -> pure (Right (fromIntegral por, sok))
-
-startServ :: (HasPierConfig e, HasLogFunc e)
+openPort :: forall e . HasLogFunc e => SockOpts -> RIO e (W.Port, Net.Socket)
+openPort SockOpts {..} = case soWhich of
+  WPSpecific x  -> insist (fromIntegral x)
+  WPChoices  xs -> loop (fromIntegral <$> xs)
+
+ where
+  loop :: [W.Port] -> RIO e (W.Port, Net.Socket)
+  loop = \case
+    [] -> do
+      logTrace "Fallback: asking the OS to give us some free port."
+      ps <- io W.openFreePort
+      logTrace (display ("Opened port " <> tshow (fst ps)))
+      pure ps
+    x : xs -> do
+      logTrace (display ("Trying to open port " <> tshow x))
+      io (tryOpen x) >>= \case
+        Left (err :: IOError) -> do
+          logWarn (display ("Failed to open port " <> tshow x))
+          logWarn (display (tshow err))
+          loop xs
+        Right ps -> do
+          logTrace (display ("Opened port " <> tshow (fst ps)))
+          pure ps
+
+  insist :: W.Port -> RIO e (W.Port, Net.Socket)
+  insist p = do
+    logTrace (display ("Opening configured port " <> tshow p))
+    io (tryOpen p) >>= \case
+      Left (err :: IOError) -> do
+        logWarn (display ("Failed to open port " <> tshow p))
+        logWarn (display (tshow err))
+        throwIO (CantOpenPort p)
+      Right ps -> do
+        logTrace (display ("Opened port " <> tshow (fst ps)))
+        pure ps
+
+  bindTo = if soLocalhost then "127.0.0.1" else "0.0.0.0"
+
+  getBindAddr :: W.Port -> IO SockAddr
+  getBindAddr por =
+    Net.getAddrInfo Nothing (Just bindTo) (Just (show por)) >>= \case
```

Comment:
Previous behavior was a bug. This actually *always* ignored the suggested ports and used whatever free ports were available.